### ISSUE TITLE
pnext: Switch to TurboText 4

### ISSUE THREAD

**Issue by eriohl** [Starcounter/Starcounter#2965]
Replace TT 3 with TT 4.

* No need to keep support for TT 3.
* No need for log upgrade since previous versions are not compatible anyway.

**Comment by eriohl** [comment]
Kernel adaption to TT4 complete. Merge pending.

**Comment by eriohl** [comment]
Need to put some effort into improving encoder and decoder performance. Especially the encoder is a lot slower with TT4.

**Comment by dan31** [comment]
CC @linkdata
My suggestion on here is that, if this doesn't show up dramatically in the tests, including some microbenchmark for TT (let's say, decrease is under 30%) and TPC-C test (unfortunately, haven't really seen yet if the test incorporate any of string ops) we can plan it into future releases and rather do more urgent items.

**Comment by linkdata** [comment]
There are microbenchmarks in the repo. IIRC, focus was on correctness first (TT3 has some serious issues), decoder speed second and encoder speed last, as according to @Starcounter-Jack, decoding is a much, much more common operation than encoding.

**Comment by eriohl** [comment]
Ok. It was my understanding that switching to TT4 because was important because of these serious issues in TT3. Thus I wanted this done before we move to **pnext** because if it is done later it will add a lot of extra work because of the format change. So in my opinion we either fix this now or kick the can into the tall grass and use TT3 for the foreseeable future.

As for performance the microbenchmarks shows TT4 being faster on compares but slower on both encoding and decoding. In the case of encoding it runs more then twice as slow. The difference here is the TT3 encoder uses a lookup table to translate the characters while TT4 does not. I experimented with introducing a lookup table for common characters and then the performance difference shrinks a lot. The rest of the performance loss I think is due to TT4 works with bytes while TT3 works with qwords, which is less easy to address. In the TPC-A benchmark without I/O the difference it smaller per thread because the lookup table yields lower IPC (although lower IPC means better use of HT so it probably scales a bit better then).

A note here that in TT3 the encoder is much faster then the decoder (because of the lookup table). I is probably true that the decoder performance is generally more important then the encoder benchmark although not in the TPC-A benchmark.

In my opinion the problem is that we are moving in the wrong direction on this. TT performance is a serious issue as it is. So we need it to be faster, not slower. If it turns out that this is hard to achieve with TT4 we have a problem. And I think it might be.

So: Should I fix this in **pnext** and hope that the performance issues works out of should I drop the project?

**Comment by dan31** [comment]
I believe we all agree we should have a single version of TT in pnext asap, before we merge it in develop. What if we do this in such order:
1) @eriohl fixes this in pnext with TT4 current track as he suggests,
2) @alemoi provides a test-run polygon to run more functional tests with broad code coverage in addition to TPC-C (perhaps in 2-3 weeks he can address it), 
3) We get confirmable feedback from these tests and work out performance issues in TT in a feedback loop with those tests.
4) Forecasting customers' feedback, "the decoder performance is generally more important then the encoder" is definitely true, since many of them expect to provide intensive read and aggregate scenarios.
@eriohl @linkdata @bigwad ?

**Comment by eriohl** [comment]
Just to be clear: You want me to merge adaption to TT4 into **pnext** asap and hope that the performance difference either doesn't matter or can be worked out later then?

**Comment by dan31** [comment]
I will also clarify: I suggest that having TT4 in pnext now and have its performance fixed soon later is better than keep pnext with TT3 and still maintain TT3 and TT4 tracks. (Having less parallel tracks will save us time and energy, and, as being said by someone as an antipattern, nothing is more permanent than the temporary). Do you agree, or you think it is a bad idea to merge adaption to TT4 into pnext asap and address performance difference soon later?

**Comment by eriohl** [comment]
I can't determine that. I just wanted to be clear what you wanted me to do.

**Comment by dan31** [comment]
I'm supporting a decision which makes most sense based on your mutual feedback. Based on this, having TT4 in pnext and have it fixed asap for performance looks the best to go. @linkdata, once @eriohl merges TT4 in pnext, how much would it take you to address performance improvement? @linkdata please confirm you agree with the scheme https://github.com/Starcounter/Starcounter/issues/2965#issuecomment-171611117.

**Comment by linkdata** [comment]
The performance difference between TT3 and TT4 in encoding/decoding is mostly due to TT3 only supporting UCS-2 and TT4 supporting full Unicode. While we can't realistically drop Unicode as a feature and limit ourselves to UCS-2, there may be some optimisations possible.

As to testing performance, we should test it in SC. Microbenchmarks are good while developing the library to give an indication if the changes one is doing is going the right way, but in the full application the code and data access patterns are very different, and what runs fast in a microbenchmark may run slow in the full application, and vice versa.

I will add the "infinity character" feature to TT4. Once that is done, it should be possible to run either TT3 or TT4 DLL's with SC, and then I hope we can do proper benchmarks by simply running SQL queries with text indices.

**Comment by eriohl** [comment]
> I will add the "infinity character" feature to TT4. Once that is done, it should be possible to run either TT3 or TT4 DLL's with SC, and then I hope we can do proper benchmarks by simply running SQL queries with text indices.

I've already done this.

**Comment by linkdata** [comment]
Awesome! :)

**Comment by linkdata** [comment]
I'll start looking at possible TT4 speedups.

**Comment by linkdata** [comment]
@eriohl : Which branch should I be working with? I see pnext-TT4, is that the correct one, with the infinity-char added?

**Comment by eriohl** [comment]
Yes.

**Comment by eriohl** [comment]
> Just to be clear: You want me to merge adaption to TT4 into pnext asap and hope that the performance difference either doesn't matter or can be worked out later then?

Done.

Note to @linkdata: pnext-TT4 branch is dropped.

**Comment by linkdata** [comment]
I just checked in a version of TT4 with a hash-based encoder that is about 50% faster in pnext-TT4. 

@eriohl Can you merge those changes into the branch you want it in, and then let me know which branch I should be working in when I start working on decoder speed?

**Comment by eriohl** [comment]
Merge what from where? Merged TT4 into pnext as described. Can't you just push you're stuff there as well?

**Comment by linkdata** [comment]
I did development in pnext-TT4 since you told me to, but you merged and removed that branch before I could push my local changes. But no problem, I'll copy my changed files from pnext-TT4 into pnext. Should be an easy merge.

**Comment by eriohl** [comment]
OK. Sorry. Should have waited a bit dropping the branch of course.

**Comment by linkdata** [comment]
Push into pnext.
